[김효인] sprint8

mozzi34 opened the pull request:
## 요구사항

#### 기본 요구사항
요구사항
기본 요구사항
공통

- [x] Github에 스프린트 미션 PR을 만들어 주세요.
- [x] Next.js를 사용해 진행합니다.

##### 자유 게시판 페이지

- [x] 게시글 목록에서 드롭다운을 사용하여 "최신 순"으로 정렬할 수 있도록 합니다.
- [x] 본인이 이전 미션에서 생성한 게시글 목록 조회 API를 활용해 GET 메서드로 데이터를 가져옵니다.
- [x] 게시글 제목에 검색어가 일부 포함되면 해당 게시글을 검색할 수 있도록 합니다.
- [x] 이미지는 디폴트 이미지로 프론트엔드에서 처리해 주세요.
- [x] 게시글 닉네임 및 좋아요 개수 역시 임의값으로 프론트엔드에서 처리해주세요.
- [x] 베스트 게시글은 최신순 3개 게시글을 요청으로 데이터를 가져와 구현해주세요.
- [x] 자유게시판 페이지에서 특정 게시글을 클릭하면 해당 게시물의 상세 페이지로 이동합니다.

##### 게시글 등록 & 수정 페이지

- [x] 각 input 필드에 정확한 placeholder 값을 입력합니다.
- [x] 모든 input 필드에 값을 입력하면 '등록' 버튼이 활성화됩니다.
- [x] 본인이 이전 미션에서 생성한 게시글 생성 API를 활용해 POST 메서드로 게시글을 등록합니다.
- [x] '등록' 버튼을 누르면 해당 게시물 상세 페이지로 이동합니다.
- [x] 게시글 수정 페이지 UI는 게시글 등록 페이지와 동일합니다.
- [x] 본인이 이전 미션에서 생성한 게시글 상세 API의 PATCH 메소드를 사용하여 게시물을 수정합니다.

##### 게시글 상세 페이지

- [x] 본인이 이전 미션에서 생성한 게시글 상세 API의 GET 메소드를 사용하여 데이터를 가져옵니다.
- [x] 본인이 이전 미션에서 생성한 게시글 상세 API의 DELETE 메소드를 사용하여 게시물을 삭제합니다.
- [x] 댓글 input에 값을 입력하면 '등록' 버튼이 활성화됩니다.
- [x] 본인이 이전 미션에서 생성한 댓글 생성 API를 활용해 POST 메소드로 댓글을 등록합니다.
- [x] 본인이 이전 미션에서 생성한 댓글 생성 API를 활용해 PATCH 메소드로 댓글을 수정합니다.
- [x] 본인이 이전 미션에서 생성한 댓글 생성 API를 활용해 DELETE 메소드로 댓글을 삭제합니다.

#### 심화 요구사항

공통

- [x] 디자인 시안에 따라 반응형 디자인을 구현합니다.
- [ ] (생략 가능) 원한다면 지금까지 진행한 모든 React 코드를 Next.js로 마이그레이션 해주세요.
- [ ] 마이그레이션에 상당한 시간이 소요될 수 있으므로 진행을 권장하지 않습니다.


## 주요 변경사항

-
-

## 스크린샷


<img width="1668" alt="스크린샷 2024-09-08 오후 9 02 42" src="https://github.com/user-attachments/assets/c5c5572c-48ca-43e9-8b60-6464e7c83aef">
<img width="1269" alt="스크린샷 2024-09-08 오후 9 02 47" src="https://github.com/user-attachments/assets/82dded6e-5c0e-4fd8-8e28-706d69bc6779">
<img width="1158" alt="스크린샷 2024-09-08 오후 9 03 13" src="https://github.com/user-attachments/assets/18e8e977-ca81-4bb2-b939-0d8ba293afcd">
<img width="1162" alt="스크린샷 2024-09-08 오후 9 03 38" src="https://github.com/user-attachments/assets/77a40605-4ab5-4bdf-890e-d4c750dc7772">


## 멘토에게

- 수정 페이지, 페이지네이션, 댓글 등록/수정/삭제, 커서 기반 페이지네이션은 아직 완성하지 못했습니다
- height 크기 조정은 일시적으로 해 놨습니다
- 여기서 파일을 기능별로 더 나눠야 할까요? (api는 나중에 한 파일로 옮길 예정입니다)
- useEffect를 이렇게 쓰는 게 맞나요?

9/15
- useRef, getServerSideProps를 처음 써 봤는데 의도대로 잘됐는지 모르겠어요
- 댓글은 커서기반으로 썼고 게시물은 다른 방법을 써 봤습니다
- 댓글 수정을 맞는 id를 찾아서 내용만 수정되게끔 했는데  댓글 수정 -> 새로 get을 통헤 전체 조회로 불러오는 게 자연스러운지 모르겟어요
`    setComments((prev) =>
        prev.map((comment) =>
          comment.id === id ? { ...comment, content: editComment } : comment
        )
      );`

mozzi34 committed:
add: Navbar

mozzi34 committed:
refactor: Navbar css

mozzi34 committed:
add: CSS for Best Articles and Posts Viewing

mozzi34 committed:
refactor: navbar 링크 추가

mozzi34 committed:
이미지 추가

mozzi34 committed:
add: 자유게시판 추가

mozzi34 committed:
refactor: freeboard link

mozzi34 committed:
add: date

mozzi34 committed:
add: footer

mozzi34 committed:
add: image

mozzi34 committed:
add: article list

mozzi34 committed:
add: 상세페이지

mozzi34 committed:
add: 검색 페이지

mozzi34 committed:
add: 작성 페이지

mozzi34 committed:
add: 기본 페이지

mozzi34 committed:
refactor: 파일 나누기

mozzi34 committed:
refactor:파일 나누기

mozzi34 committed:
코드잇 기본 파일

mozzi34 committed:
fix: comment date

mozzi34 committed:
fix: comment time

mozzi34 committed:
add: 댓글 수정/삭제/추가 기능 추가

mozzi34 committed:
refactor: 검색 페이지 css 수정

mozzi34 committed:
refactor: console.log 삭제

mozzi34 committed:
add: json

ehj0128 commented on components/ArticleDetail/ArticleDetailInfo.js:
```diff
@@ -0,0 +1,70 @@
+  function handleDelete() {
+    deleteArticle(id);
+    router.push('/freeboard');
+  }
```
함수형태를 맞추면 좋을 것 같아요!

그리고 나중에는 최적화를 위해서 useCallback도 한번 사용해보세요

ehj0128 commented on components/ArticleDetail/ArticleDetailInfo.js:
```diff
@@ -0,0 +1,70 @@
+          />
+          {openOptions && (
+            <div className={styles.dropDown}>
+              <Link href={`/article/patch/${id}`} className={styles.link}>
```
url에 patch가 들어가는게 좀 어색합니다. update나 edit같은 것을 추천합니다. github도 readme 수정하려고 하면 url에 edit을 쓰고 있네요

ehj0128 commented on components/ArticleDetail/ArticleDetailInfo.js:
```diff
@@ -0,0 +1,70 @@
+              <Link href={`/article/patch/${id}`} className={styles.link}>
+                <div className={styles.dropDownText}>수정하기</div>
+              </Link>
+              <div className={styles.dropDownDelete} onClick={handleDelete}>
```
삭제같은 경우는 유저에게 한번도 확인하는 케이스도 있는 것 같아요

ehj0128 commented on components/ArticleDetail/CommentList.js:
```diff
@@ -0,0 +1,124 @@
+import styles from '@/styles/Comment.module.css';
+import { useState } from 'react';
+
+function CreateDate({ createDate }) {
```
CreateDate 보다 dateFormat 같은걸로 네이밍 하면 좋을 것 같습니다.

ehj0128 commented on components/ArticleDetail/CommentList.js:
```diff
@@ -0,0 +1,124 @@
+                            onClick={handleDelete}
+                          >
+                            삭제하기
+                          </div>
```
저는 버튼별로 컴포넌트 만들기를 추천합니다. 

여기서 스타일도 같기 때문에 onClickHandler랑 label(수정하기 ,삭제하기) 등만 porps로 보내주면 될 것 같아요 다른 드롭다운버튼에서도 활용할 수 있을 것 같고요

ehj0128 commented on components/ArticleDetail/Comments.js:
```diff
@@ -0,0 +1,110 @@
+        className={styles.inputComment}
+      />
+      <button
+        disabled={!submit}
```
변수이름을 submit보다는 canSubmit 이라던지 확인을 나타낼 수 있는 변수가 좋은것 같아요 ture/false는 보통 is can들을 많이 붙이고 이런 동사형은 바로 사용하지 않는 것 같아요

ehj0128 commented on components/ArticleDetail/EditComment.js:
```diff
@@ -0,0 +1,69 @@
+  }
+
+  useEffect(() => {
+    if (editComment) {
```
굳이 editComment와 submit을 분리할 이유가 있을 까요?

const submit = !!editComment 아닌가요? 

editComment가 변화하면 그것에 따라 disable도 처리할 수 있을 것 같아요

mozzi34 replied:
따로 관리하는 게 더 나을 것 같다고 생각했는데 분리 안 해도 되는 거면

const submit = !!editComment 로 안 하고 아예 삭제해도 괜찮을까요?

```
        <button
          disabled={!editComment}
          className={editComment ? styles.submitBtn : styles.btn}
          onClick={handleSubmit}
          type='button'
        >
```

ehj0128 commented on components/Footer.js:
```diff
@@ -0,0 +1,34 @@
+            <Image src={facebookLogo} alt='코드잇 로고' />
+            <Image src={youtuveLogo} alt='코드잇 로고' />
+            <Image src={instaLogo} alt='코드잇 로고' />
+            <Image src={twitterLogo} alt='코드잇 로고' />
```
로고가 다 달라요

ehj0128 commented on components/FreeBoard/ArticleList.js:
```diff
@@ -0,0 +1,41 @@
+        )}
+      </div>
+    </>
+  );
```
저라면 early return 으로 아래와 같이 작성할 것 같아요.

```
if (articles.length === 0) {
  return <div className={styles.articleList}>
               <span>둘러볼 게시글이 없습니다</span>
             </div>
}

return <있을 때 />

ehj0128 commented on components/FreeBoard/SearchForm.js:
```diff
@@ -0,0 +1,38 @@
+      router.push('/freeboard');
+      return;
+    }
+    router.push(`/freeboard/search?keyword=${value}`);
```
검색어가 붙었다고 url이 달라지는건 좀 어색한 것 같습니다. freeboard?keyword=value 이렇게 바로 오는것이 더 일반적입니다.

ehj0128 commented on pages/article/[id].js:
```diff
@@ -0,0 +1,69 @@
+      );
+      const nextArticle = res.data;
+      setArticle(nextArticle);
+      setArticleId(nextArticle.id);
```
nextArticle을 가져오기위한 커서인가요? 맞다면 굳이 useState로 할 필요 없고 useRef를 사용하면 어떨까요? useState는 변화하면 렌더링을 일으켜서 별로 좋은 것 같진 않아요

ehj0128 replied:
근데 왜 nextArticle이죠?

mozzi34 replied:
게시물 전체 조회할 때 nextArticle 사용한 걸 그대로 갖고 왔어요 수정 예정입니다!

ehj0128 commented on pages/article/patch/[id].js:
```diff
@@ -0,0 +1,97 @@
+  };
+
+  useEffect(() => {
+    setSubmit(titleValue.trim() !== '' && contentValue.trim() !== '');
```
이런 trim작업은 onChange에서 하면 좋을 것 같아요

ehj0128 commented on pages/article/patch/[id].js:
```diff
@@ -0,0 +1,97 @@
+          value={titleValue}
+          onChange={titleChange}
+          className={styles.input}
+        ></input>
```
태그 내부에 내용이 없으면
```
<input
  placeholder='제목을 입력하세요'
  value={titleValue}
  onChange={titleChange}
  className={styles.input}
/>
```
로 사용합니다.

ehj0128 reviewed (commented):
- 여기서 파일을 기능별로 더 나눠야 할까요? (api는 나중에 한 파일로 옮길 예정입니다)
  - 회사마다 컨벤션은 다르지만 page에 있는 폴더는 간단할 수록 좋은 것 같아요 명확하게 기능에 따라 분리되고 한눈에 들어와야 좋은 것 같습니다. 
- useEffect를 이렇게 쓰는 게 맞나요?
   - 이렇게가 어떻게 인가요? async호출 하는걸 말씀하시는 건가요? 그렇다면 맞습니다. 공식문서에서는 useEffect 안에 async 함수를 정의해서 사용하곤 합니다. 

```import { useState, useEffect } from 'react';
import { fetchBio } from './api.js';

export default function Page() {
  const [person, setPerson] = useState('Alice');
  const [bio, setBio] = useState(null);
  useEffect(() => {
    async function startFetching() {
      setBio(null);
      const result = await fetchBio(person);
      if (!ignore) {
        setBio(result);
      }
    }

    let ignore = false;
    startFetching();
    return () => {
      ignore = true;
    }
  }, [person]);

  return (
    <>
      <select value={person} onChange={e => {
        setPerson(e.target.value);
      }}>
        <option value="Alice">Alice</option>
        <option value="Bob">Bob</option>
        <option value="Taylor">Taylor</option>
      </select>
      <hr />
      <p><i>{bio ?? 'Loading...'}</i></p>
    </>
  );
}
```

https://react.dev/reference/react/useEffect

그러나... react-query를 사용하게 되면 api 호출을 useEffect에서 사용할일은 거의 없을 겁니다!!

mozzi34 committed:
refactor(mentor): CreateDate 파일 이름 DateFormat으로  변경

mozzi34 committed:
refactor(mentor): submit state를  canSubmit state로 변경

mozzi34 committed:
refactor(mentor): footer sns logo 위치 변경

mozzi34 committed:
refactor(mentor): sns 로고 alt 변경

mozzi34 committed:
refactor(mentor): ArticleList early return

mozzi34 committed:
refactor(mentor): path[id]
- 게시물수정 Url path에서 edit로 수정
- setSubmit을 onChange로 업데이트 수정

mozzi34 committed:
refactor: searchPage url 변경
- serarch?keyword 에서 -?keyword로 변경
- 검색 시 게시글 타이틀에서 검색 결과로 변경
- 자유게시판 클릭 시 input창 키워드 지우기

mozzi34 committed:
refactor(mentor): [ArticleDetailInfo.js] 함수 형태 수정

mozzi34 committed:
refactor(mentor): DropDown 컴퍼넌트 추가

mozzi34 committed:
refactor: css 스타일 및 오류
- 게시글 수정 후 저장되지 않은 게시물로 이동하는 오류 해결
- 삭제 시 페이지 이동 수정
- nav margin-top 추가
- 베스트게시글 조회 시 제목 잘리는 경우 수정

mozzi34 committed:
refactor(mentor):add  button Component

mozzi34 committed:
refactor: post, edit Component 분리

mozzi34 committed:
delete: 임시 회원가입 페이지 삭제

mozzi34 committed:
refactor: post, edit 기능 분리

mozzi34 committed:
add: favicon.ico

mozzi34 committed:
refactor: api 주소 수정

mozzi34 committed:
refactor: comments 커서 기반 페이지네이션 추가

mozzi34 committed:
add: comments cursor pagination

mozzi34 committed:
refactor: 게시글 키워드 검색 에러

mozzi34 committed:
refactor: comments 수정 안 되는 오류로 인한 수정

mozzi34 committed:
refactor: 게시물 api 분리, SSR

mozzi34 committed:
refactor: comment API

mozzi34 committed:
refactor: 게시물 get 파일 분리

mozzi34 committed:
refactor: 댓글 api 분리

mozzi34 committed:
refactor: 게시물 api 분리

mozzi34 committed:
refactor: 게시물 삭제 알림창 추가

mozzi34 committed:
refactor: 댓글 api 수정

mozzi34 committed:
refactor: 게시물 api 수정

mozzi34 committed:
refactor: 댓글 api 수정

mozzi34 committed:
refactor: 반응형 웹페이지

mozzi34 committed:
refactor: 모바일화면 nav 수정

mozzi34 committed:
refactor: 모바일 css 수정

mozzi34 committed:
refactor(mentor): 댓글 시간 이름 수정

mozzi34 commented on components/ArticleDetail/Comments.js:
```diff
@@ -0,0 +1,95 @@
+    setCommentsList([]);
+  }, [articleId, canEdit]);
+
+  useEffect(() => {
```
스크롤을 여기 말고 useCommets로 옮기는 게 나았을까요?

ehj0128 replied:
스크롤 관련 Hook을 따로 만들어도 괜찮을것 같아요 만약 다른 곳에서 무한스크롤을 또 사용하려면 이용할수 있으니까요!
그리고 스크롤 관련 이벤트는 저번에 input에서 바로바로 검색되는 api 요청을 막기위해 썼던 throttle이나 debounce 적용해 보는 것 도전해봐도 괜찮을 것 같습니다.

mozzi34 committed:
refactor(mentor): cursorId useRef 사용

ehj0128 commented on .gitignore:
```diff
@@ -35,6 +35,6 @@ yarn-error.log*
 next-env.d.ts
 
-*.json
+package-lock.json
```
package-lock.json 을 왜 추가하셨을까요 ? ㅎㅎ package-lock.json의 역할도 study 해보셔요

https://hyunjun19.github.io/2018/03/23/package-lock-why-need/

mozzi34 replied:
감사합니다 package-lock.json 역할을 공부해 보겠습니다....!

ehj0128 commented on components/ArticleDetail/ArticleDetailInfo.js:
```diff
@@ -1,58 +1,73 @@
+  }, []);
+
+  const buttonEvent = async () => {
+    if (confirm('정말 삭제하시겠습니까??') == true) {
```
=== 사용!!!

ehj0128 commented on components/ArticleDetail/ArticleDetailInfo.js:
```diff
@@ -1,58 +1,73 @@
+                onClickHandler: handleEdit,
+                label: '수정하기',
+              }}
+              secondAction={{
```
3번째 4번째 계속 생기나요? ㅋㅋㅋㅋ

mozzi34 replied:
ㅋㅋㅋㅋㅋ 계속 드롭다운이 늘어난다면 ... 생길 수도 있을 것 같아요.... 다른 이름으로 했어야 했을까요? ㅠㅠ ㅋㅋㅋㅋ

ehj0128 commented on components/ArticleDetail/CommentList.js:
```diff
@@ -45,49 +47,60 @@ export default function CommentList({
     setEditId(e);
   }
 
+  if (comments.length === 0) {
```
분리한거 좋아요!! early return 으로 코드가 좀 더 깔끔해졌네요.

ehj0128 commented on components/ArticleDetail/Comments.js:
```diff
@@ -1,85 +1,75 @@
+    loading,
+    cursorIdRef,
+    hasMore,
+  } = useComments({
```
필요한 함수 hook으로 정리한거 너무 좋습니다!

ehj0128 commented on components/FreeBoard/BestArticleList.js:
```diff
@@ -1,17 +1,45 @@
+export default function BestArticleList({ articles }) {
+  const [bestArticles, setBestArticles] = useState(articles);
+
+  useEffect(() => {
```
스크린 사이즈를 통해서 device를 유추하는 hook을 만들어 사용하면 좀 더 좋을 것 같아요.

ehj0128 commented:
```
댓글 수정을 맞는 id를 찾아서 내용만 수정되게끔 했는데 댓글 수정 -> 새로 get을 통해 전체 조회로 불러오는 게 자연스러운지 모르겠어요
    setComments((prev) => prev.map((comment) => comment.id === id ? { ...comment, content: editComment } : comment ) );
```
질문의 의도를 모르겠지만 GET api 요청없이 state만 변경하신것 맞나요? 맞다면 이렇게 하기도 합니다.

아마 나중에 react-query 를 사용하게 된다면 데이터를 새로 가져올 수 도 있고 상태만 변화시킬수도 있고 선택할 수 있습니다.

mozzi34 committed:
refactor: package-lock.json 추가

mozzi34 committed:
refactor(mentor): === 사용

mozzi34 committed:
refactor(mentor): 댓글 스크롤 hook 생성

mozzi34 committed:
refactor(mentor): add Device hook

mozzi34 committed:
refactor: bestArticle, freeboaed 반응형 웹페이지 css 수정
- bestArticles: 가운데 정렬 추가
- freeboard: 검색창 100% 추가

mozzi34 committed:
refactor(mentor): scroll throttle 추가

mozzi34 committed:
add: README
